gammapy/maps/hpxnd.py

```diff
@@ -278,31 +271,28 @@ def pad(self, pad_width, mode='constant', cval=0.0, order=1):
 
     def crop(self, crop_width):
         geom = self.geom.crop(crop_width)
-        map_out = self.__class__(geom, meta=copy.deepcopy(self.meta), unit=self.unit)
+        map_out = self.clone(geom=geom, data=None)
         map_out.coadd(self)
         return map_out
 
     def upsample(self, factor, preserve_counts=True):
-
-        map_out = self.__class__(self.geom.upsample(factor),
-                                 meta=copy.deepcopy(self.meta), unit=self.unit)
-        coords = map_out.geom.get_coord(flat=True)
-        vals = self.get_by_coord(coords)
-        m = np.isfinite(vals)
-        map_out.fill_by_coord([c[m] for c in coords], vals[m])
+        geom = self.geom.upsample(factor)
+        coords = geom.get_coord()
+        data = self.get_by_coord(coords)
 
         if preserve_counts:
-            map_out.data /= factor ** 2
+            data /= factor ** 2
 
-        return map_out
+        return self.clone(geom=geom, data=data)
 
     def downsample(self, factor, preserve_counts=True):
+        geom = self.geom.downsample(factor)
+        coords = self.geom.get_coord()
+        vals = self.get_by_coord(coords)
```

Comment:
Isn't it slower to use `get_by_coord` rather than `get_by_idx` used in the initial implementation?